File: src/0081_search_in_rotated_sorted_array_ii.js
function search(nums, target) {
    let left = 0;
    let right = nums.length - 1;
    while (left <= right) {
        const mid = left + Math.floor((right - left) / 2);
        if (nums[mid] === target) return true;
        if (nums[left] < nums[mid]) {
            if (nums[left] <= target && target < nums[mid]) right = mid - 1;
            else left = mid + 1;
        } else if (nums[left] > nums[mid]) {
            if (nums[mid] < target && target <= nums[right]) left = mid + 1;
            else right = mid - 1;
        } else {
            left++;
        }
    }
    return false;
}

module.exports = { search };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(search([2, 5, 6, 0, 0, 1, 2], 0), true);
    assert.deepStrictEqual(search([2, 5, 6, 0, 0, 1, 2], 3), false);
}

File: src/0081_search_in_rotated_sorted_array_ii.rs
pub fn search(nums: Vec<i32>, target: i32) -> bool {
    let mut left = 0_usize;
    let mut right = nums.len().saturating_sub(1);

    while left <= right && !nums.is_empty() {
        let mid = left + (right - left) / 2;
        if nums[mid] == target {
            return true;
        }

        if nums[left] < nums[mid] {
            if nums[left] <= target && target < nums[mid] {
                if mid == 0 {
                    break;
                }
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        } else if nums[left] > nums[mid] {
            if nums[mid] < target && target <= nums[right] {
                left = mid + 1;
            } else if mid == 0 {
                break;
            } else {
                right = mid - 1;
            }
        } else {
            left += 1;
        }
    }

    false
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(search(vec![2, 5, 6, 0, 0, 1, 2], 0), true);
        assert_eq!(search(vec![2, 5, 6, 0, 0, 1, 2], 3), false);
    }
}
